### ISSUE TITLE
Rule request: dictionary-interfaces

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [microsoft/tslint-microsoft-contrib#334]
Some teams have something like:

```typescript
interface IDictionary<T> {
    [i: string]: T;
}
```

There's no need to make other interfaces that do the same thing when this one is available.

```typescript
// Should really just be replaced with IDictionary<string>,
// or interface IPeopleByName extends IDictionary<string>,
// or type IPeopleByName = IDictionary<string>;
interface IPeopleByName {
    [name: string]: string;
}
```

How about a rule that bans anything that boils down to the above `IDictionary`?

**Comment by karol-majewski** [comment]
I can take a stab at this one, but I have a few questions first.

1. I presume the desired identifier name should be configurable — some users will use `Dictionary`, some `IDictionary` while others may use `RecordOf`. Should there by only one name allowed, or should the rule accept an array of names?
1. The predicate is for the interface or type to have _just_ the index signature. Should all these be considered violations?

```ts
interface DictionaryInterface<T> {
    [index: string]: T;
}

type DictionaryType<T> = {
    [index: string]: T;
};

type DictionaryTypeWithoutTypeParameter = {
    [index: string]: any;
};

interface DictionaryInterfaceWithoutTypeParameter {
    [index: string]: any;
}
```

And also this one mentioned in the original post. Should such a statement be always allowed?

```ts
interface ExtendsDictionaryInterface<T> extends DictionaryInterface<T> {}
```

**Comment by JoshuaKGoldberg** [comment]
Super, looking forward to your PR @karol-majewski! 

1. Agreed, there should be a configurable list of names that may be taken in. IMO `Dictionary` and `IDictionary` are fine defaults.
2. Agreed, your `DictionaryInterface<T>`, `DictionaryType<T>` , `DictionaryTypeWIthoutTypeParameter`, and `DictionaryInterfaceWithoutTypeParameter` would be violations. `ExtendsDictionaryInterface<T>` would not be a violation.

**Comment by JoshuaKGoldberg** [comment]
Super, looking forward to your PR @karol-majewski! 

1. Agreed, there should be a configurable list of names that may be taken in. IMO `Dictionary` and `IDictionary` are fine defaults.
2. Agreed, your `DictionaryInterface<T>`, `DictionaryType<T>` , `DictionaryTypeWIthoutTypeParameter`, and `DictionaryInterfaceWithoutTypeParameter` would be violations. `ExtendsDictionaryInterface<T>` would not be a violation.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated**; therefore, it and tslint-microsoft-contrib are no longer accepting pull requests for major new changes or features. See #4534. 😱

If you'd like to see this change implemented, you have two choices:
* **Recommended:** Check if this is available in [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io) ✅
* _Not Recommended: Fork TSLint or tslint-microsoft-contrib locally_ 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
